### ISSUE TITLE
Swaylock not getting through pam check

### ISSUE THREAD

**Issue by x10an14** [NixOS/nixpkgs#158025]
### Describe the bug
My `swaylock` is unable to unlock my password/sway DE.

### Steps To Reproduce
1. `sudo nixos-rebuild switch --flake https://git.sr.ht/~x10an14/nixos-config/commit/60d5555977ebc0cd1e6a52b1394fd362f5ba245b#`
1. `home-manager switch --flake https://git.sr.ht/~x10an14/home-manager-config/commit/7d0fcfe9922a7772422d697fc2d70f79901c8c30`
1. Log into sway
1. `swaylock`

### Expected behavior
Correct password input into `swaylock` to unlock and return user to `sway`.

### Additional context
I install swaylock/swaylock-effects through home-manager.
Due to https://github.com/NixOS/nixpkgs/issues/143365 I have the below enabled in my flaked nixos configuration (separate/independent flake from my hm config).
```nix
security.pam.services.swaylock = {};
```

This is what appears in the output of `sway` itself (1x per attempted unlock):
```
[pam.c:101] pam_authenticate failed: invalid credentials
```
---
As mentioned in this issue thread (https://github.com/NixOS/nixpkgs/issues/157112#issuecomment-1024250758), I had a similar issue 6 days ago.
But back then that fix referenced worked for me. And has worked until today/yesterday.

It doesn't work, even after updating to latest releases of `nixos-unstable` as shown in above flakes, and explained here: https://github.com/NixOS/nixpkgs/issues/157112#issuecomment-1029282194

### Notify maintainers
@gnxlxnxx @primeos 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.16.4, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.6.0`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by Synthetica9** [comment]
Ran into the same issue. My temporary fix is to hold swaylock back to the commit before sway 1.7 was introduced, 2e719d1cdab9e81750c19425a1f5c7678b5e73ad (I also use 1.6 because of the chromium regression.)

**Comment by Synthetica9** [comment]
Wrote a test (#158027), but turns out it's more specific than I thought (that just runs on master). More investigation needed.

**Comment by x10an14** [comment]
Let me know if and how I can help!

**Comment by Synthetica9** [comment]
I'm just gonna let a bisect run overnight, see what we find. I'll report back in the morning with the result.

**Comment by Synthetica9** [comment]
Bisect isolated 62e6374d5e1cecfdedd702592f09bdbb87249e5d as the first bad commit, which seems plausible. Currently testing that hypothesis.

**Comment by Synthetica9** [comment]
Nope that was unrelated build failure... Sadly `nix-bisect --failure-line` is also broken... 🙃

**Comment by x10an14** [comment]
Oh well x) Still here if I can assist in any way!

**Comment by x10an14** [comment]
Missclick...

**Comment by primeos** [comment]
I cannot reproduce this (tested via commit 554d2d8aa25b6e583575459c297ec23750adb6cb / nixos-unstable). Maybe it is related to Home Manager? It seems like you install Swaylock via NixOS but Sway via Home Manager? That should be fine but maybe there are some impurities (we usually have those mainly through /run/opengl-driver/ though but it could theoretically cause unrelated issues).

On an unrelated note: IMO Home Manager users should set `programs.sway.enable = true;` (the rest can still be configured through Home Manager). Unfortunately that probably still isn't documented (IIRC I requested that once)...

@Synthetica9 are you by any chance using that Home Manager approach too?

**Comment by Synthetica9** [comment]
No I use configuration.nix

**Comment by x10an14** [comment]
Any suggestions (however temporary) of how to remediate this?

If its enough to use an older version of Swaylock (for now) I'd happily do that.
(Need to learn how to use overlays like that first though).

**Comment by jamesluke** [comment]
I may be betraying my lack of knowledge about PAM, but is this down to running the wrong `unix_chkpwd`?

Ran `strace -f --trace=execve swaylock -d`, entered my password, then waited (for an oddly long time) for verification to fail. Output includes the line:

```[pid  1850] execve("/nix/store/g55rkqdbnq1rfmfcg2ldp29m55answ31-linux-pam-1.5.1/sbin/unix_chkpwd", ["/nix/store/g55rkqdbnq1rfmfcg2ldp"..., "<username removed>", "nonull"], 0x7f8803159040 /* 0 vars */) = 0```

Shouldn't this `execve()` be passed the setuid wrapper in `/run/wrappers/bin/unix_chkpwd`, rather than the non-setuid binary in `/nix/store`?

**Comment by primeos** [comment]
@jamesluke good catch, that should indeed be a problem. However, when using `strace` that part should unfortunately always fail. I forgot about the details but IIRC the tl;dr is that `strace` is based on the `ptrace` syscall which allows for much more than just tracing (e.g., reading from and writing to memory of the traced program). Therefore one isn't allowed to trace privileged programs as that would allow for easy privilege escalation. Anyway, I gave this a quick test on my system and when I launch Swaylock via strace it also doesn't work and doesn't use `/run/wrappers/bin/unix_chkpwd` (which comes first in `$PATH`). But without `strace` it works as expected.

**Comment by x10an14** [comment]
> @primeos But without `strace` it works as expected.

Am I understanding you correctly that with
```
programs.sway.enable = true;
```
It works for you?

I've tried with that setting too now(in addition to my home-manager settings), and I still can't log in =(

**Comment by primeos** [comment]
Yes, it works on both my main system and my test/old system. The new VM test (https://github.com/NixOS/nixpkgs/pull/158027) does also work.

It seems like there must be some other settings that can break things.

> I've tried with that setting too now(in addition to my home-manager settings), and I still can't log in =(

Thanks, it was worth a try and good to know that this doesn't fix this particular issue either.

**Comment by x10an14** [comment]
I think I may have narrowed it down, though not 100% sure.

I can lock (and unlock) with Swaylock installed from NixOS (`/nix/store/3dhs4kbd15z4ngxnacrx99kfcywnjgc1-swaylock-1.5/bin/swaylock`).
But I cannot do so with [Swaylock effects](https://search.nixos.org/packages?channel=21.11&show=swaylock-effects&from=0&size=50&sort=relevance&type=packages&query=swaylock-effects) installed via home-manager.

I've tried both by just executing their `/nix/store` binaries manually.

Does anyone here (besides @gnxlxnxx who hasn't responded yet) use Swaylock effects?
How could one make a test for this, like the [one you mentioned](https://github.com/NixOS/nixpkgs/pull/158027) @primeos?

**Comment by x10an14** [comment]
I think I may have narrowed it down, though not 100% sure.

I can lock (and unlock) with Swaylock installed with my nixos-config, as long as this config exists on my system:
```
programs.sway.enable = true;
```
I know it is this one when I test because I use `/run/current-system/sw/bin/swaylock` to when I run/test it.

Doesn't seem to matter if it's [swaylock-effects](https://search.nixos.org/packages?channel=21.11&show=swaylock-effects&from=0&size=50&sort=relevance&type=packages&query=swaylock-effects) or default `swaylock`.
I tested that supposition like so:
```
# nixos config:
  programs.sway = {
    enable = true;
    extraPackages = with pkgs; [swaylock-effects];
  };
# bash output (`--clock` option doesn't exist in normal swaylock):
 $ /run/current-system/sw/bin/swaylock --help 2>&1 | grep clock
  --clock                          Show time and date.
```

If I _only_ keep the abovementioned `programs.sway.enable = true;` config, and I use the home-manager installed `~/.nix-profile/bin/swaylock`, it doesn't work.

So my research has given me two questions:
1. How could one make a test for this, like the [one you mentioned](https://github.com/NixOS/nixpkgs/pull/158027) @primeos?
2. Should I close this issue and re-open it with Home-Manager?

I have no idea where to go from here.

**Comment by primeos** [comment]
> 1. How could one make a test for this, like the https://github.com/NixOS/nixpkgs/pull/158027 @primeos?

Home Manager is separate from Nixpkgs / NixOS so I don't see that happening (technically/theoretically, it should be possible with enough effort).

> 2. Should I close this issue and re-open it with Home-Manager?

Yes, IMO it's an Home Manager issue then. Or at least it should be first further analyzed/debugged there (I'm not familiar with it so I cannot help with that anyway).

> I have no idea where to go from here.

Someone who's affected by this issue needs to figure out what's going wrong (and if Nixpkgs changes are required).

**Comment by zseri** [comment]
with c7d68f57c842d56046d31688dee4e0956bc90438, I also ran into this problem, but I globally installed swaylock (e.g. `programs.sway.enable=true;`); a caveat however is that I use swaylock via swayidle, which gets started via user' systemd, and messes with the `PATH` variable...

**Comment by colemickens** [comment]
I haven't read this thread closely but I suspect this has to do with how we're importing the environment. Since either in the default config and/or the HM module, we use `systemctl --user import-environment` instead of `dbus-update-activation-environment --systemd`. I have this and a bunch of other little cleanups queued up.

(Also, one would want to ensure that they either the nixos program option, or that if they use the HM module directly, they somehow also enable the pam wrapper for swaylock so it can properly do its job).

**Comment by coreyoconnor** [comment]
I'm running into this as well ( https://github.com/coreyoconnor/sway-gnome )

@colemickens should `systemctl --user import-evnrionment` or `dbus-update-activation-environment` be used? Or are both needed for specific imports? I'm entirely ignorant of both. `sway-gnome` uses both (copy-pasted without much thought) throughout.

**Comment by coreyoconnor** [comment]
I still don't understand dbus. However, I fixed the issue I was encountering. My config was missing the following:

```
security.pam.services.swaylock = {};
```

Which, once added, corrected the issue of swaylock rejecting a valid password.

**Comment by ManoftheSea** [comment]
I'm running into this.  With nixos config, I have "programs.sway.enable = true", but I don't get a "/etc/pam.d/swaylock" configuration.  I am also using home manager, with "wayland.windowManager.sway".

I attempted to add "security.pam.services.swaylock = {}" to nixos configuration as mentioned above, but I do not see a change in /etc/pam.d nor swaylock.  Copying /etc/pam.d/login to /etc/pam.d/swaylock allows success.  I believe this is because swaylock ends up using "other", which is just set to "warn and deny"

**Comment by alejandro-angulo** [comment]
I ran into this too and setting `security.pam.services.swaylock = {}` like @coreyoconnor suggested worked :tada: 

Should this be introduced to the `security,pam` module? I see something similar is already being done for a few other programs https://github.com/NixOS/nixpkgs/blob/d10e69c86bcd415f5c4347d3c6b981241e913465/nixos/modules/security/pam.nix#L1281-L1286 but there's a comment which makes me feel like we don't want to keep doing that.

This works when enabling `programs.sway` because this is set in the nix file https://github.com/NixOS/nixpkgs/blob/d10e69c86bcd415f5c4347d3c6b981241e913465/nixos/modules/programs/sway.nix#L151

Should there instead be a new `programs.swaylock` module that creates the pam service?

I'm happy to create a PR for this, but I'm not sure what the best way to handle this might be.

**Comment by dwarf-king-hreidmar** [comment]
Thanks! I had the same issue because i use hyprland but still use swaylock for locking my system. I'm pretty new to nix but is this something that the swaylock nixos package maintainer should be notified of to fix in the package or does this have to remain a security.pam setting?

**Comment by jlamur** [comment]
Here is a fix for non-NixOS hosts.

You probably need to install a PAM configuration at `/etc/pam.d/swaylock` and setup the `/run/wrappers/bin/unix_chkpwd` suid binary.

One solution for the PAM configuration could be to copy the default configuration from the nixpkgs:
```bash
cp $(nix path-info nixpkgs#swaylock.out)/etc/pam.d/swaylock /etc/pam.d/swaylock
```

Some distributions will mount `/run` with the `nosuid` flag. You should probably simply mount a `tmpfs` at `/run/wrappers` without the `nosuid` option, with eg.:
```bash
echo "tmpfs /run/wrappers tmpfs defaults,nodev,noatime,mode=755 0 0" >> /etc/fstab
```

Then you can create a oneshot systemd unit to create a symlink from the `/sbin/unix_chkpwd` binary to `/run/wrappers/bin`:
```bash
cat <<EOF > /usr/local/lib/systemd/system/install-unix_chkpwd-wrapper.service
[Unit]
After=run-wrappers.mount
Wants=run-wrappers.mount

[Service]
Type=oneshot
ExecStart=/usr/bin/ln -sf /sbin/unix_chkpwd /run/wrappers/bin/unix_chkpwd

[Install]
WantedBy=multi-user.target
EOF
```
